SearchKit - Add "in-place edit" feature

colemanw opened the pull request:
Overview
----------------------------------------
Adds a new-fangled "in-place edit" feature to search displays. It's all the rage with the kids.

As an added bonus, this gives the option to set links to open in a new tab or a dialog.

![image](https://user-images.githubusercontent.com/2874912/110497670-ad0a2280-80c4-11eb-82d4-14e13ef18560.png)

colemanw committed:
SearchKit - Add "in-place edit" feature

colemanw committed:
SearchKit - Additional link settings for opening in new tab or popup

eileenmcnaughton commented:
Just noting that we discussed this & we think that unless someone has 'edit all contacts' they should not be able to edit contacts - for other entities it might be slightly trickier to find the right permission

dptarrant commented:
Seems to work but very difficult and confusing to work through all of the options - menus, links, buttons, etc.     If there is any way to put the styles icons, etc. on another page or popup or something that would allow access but not overload the primary features.   Will test more once I can install and use SearchKit with more data variations

dptarrant commented:
Display layout options need better organization - see image

e.g. Allows buttons, (or menus) to be added more than once.   The layout of the "add" items is according to when added rather than by category.  Note that gender is not presented in group of other fields but after/between menus/buttons
![Screenshot 2021-03-13 134542](https://user-images.githubusercontent.com/10925508/111040735-786ad380-8402-11eb-976a-210ff7314eed.png)

![image](https://user-images.githubusercontent.com/10925508/111040706-5709e780-8402-11eb-8599-2650bb235311.png)

colemanw commented:
@dptarrant the links/buttons/menu are not intended to be used more than once in a display, but I wouldn't want to entirely prevent that configuration as there could be some legitimate use for it somewhere.

In regards to the ordering of display columns, please note that they can be rearranged with drag-n-drop.

But in regards to this PR, I would note that those are unrelated to this feature, which simply adds an "In-place edit" checkbox to display fields and does not affect link/button/menu items.

dptarrant commented:
OK, please note that it is really hard to test if you don't know how it is intended to work.

eileenmcnaughton commented:
On the popup it's popping up without all the styling - is that the same for you?

![image](https://user-images.githubusercontent.com/336308/111059554-5f921900-84fb-11eb-85aa-488a6605dd53.png)

eileenmcnaughton commented:
@colemanw this seems OK for my admin user (the styling issue was the only thing I noticed)

However, when I created a less permisisoned user to test it with I get a js error looking at it - the user has
- access CiviCRM
- view all contacts
- access ajax api
And it failed with a js error - which I don't think is suitably 'graceful'

![image](https://user-images.githubusercontent.com/336308/111059678-4d64aa80-84fc-11eb-86b8-b12043c49c5d.png)


/civicrm/admin/search#/create/Participant?params=%7B%22version%22:4,%22select%22:%5B%22id%22,%22Participant_Contact_contact_id_01.display_name%22,%22Participant_Contact_contact_id_01.gender_id:label%22,%22status_id:label%22%5D,%22orderBy%22:%5B%5D,%22where%22:%5B%5D,%22groupBy%22:%5B%5D,%22join%22:%5B%5B%22Contact%20AS%20Participant_Contact_contact_id_01%22,false,%5B%22contact_id%22,%22%3D%22,%22Participant_Contact_contact_id_01.id%22%5D%5D%5D,%22having%22:%5B%5D%7D

After adding 'view event participants' and 'access CiviEvent' (both are required) the view loaded and the gender field showed as editable but gave an error if I tried to edit it

I *think* you have restricted this to the contact entity - in which case the short cut to getting this merged and punting out bigger permissions decisions would be to only show edit links to contacts with 'edit all contacts' for now

colemanw commented:
> On the popup it's popping up without all the styling - is that the same for you?

@eileenmcnaughton the "Contact Summary" screen cannot be opened in a popup... there's too much stuff on there that expects to be a full page. Other links work ok as popups though.

eileenmcnaughton commented:
@colemanw we perhaps should block it being attempted in the UI then (or at least warn)

colemanw committed:
SearchKit - add draggable arrows when configuring display columns

colemanw commented:
I've updated this PR to check permissions before presenting fields as editable. It's imperfect but adequate for now, as they check high-level permissions not fine-grained ACLs. Maybe someday.

Also updated the UI based on @dptarrant's feedback. Added draggable arrows to indicate that display columns can be rearranged, and removed the distracting color stripes for links/buttons/menus.

eileenmcnaughton commented on ext/search/Civi/Api4/Action/SearchDisplay/Get.php:
```diff
@@ -0,0 +1,66 @@
+      case 'Contact':
+        return \CRM_Core_Permission::check('edit all contacts');
+
+      // FIXME: Other entities have the opposite problem, and may be under-restrictive, as the user may have permission
```
@colemanw so I'm not fully following here - can we flesh this comment out a bit more with an example / why the option chosen addresses it- obviously we want to err on the side of being restrictive

colemanw replied:
@eileenmcnaughton basically the `'default':` case here is checking the gate-keeper permission for `api.update`. For example, the gate-keeper permission for `Campaign::update` is `('administer CiviCampaign' OR 'manage campaign')`.

But the `Contact::update` has no gate-keeper permission, because all actions on that entity are managed by ACLs. That's why I added a separate case for it. (and FYI no this new edit-in-place feature is not limited to contacts, it's limited to non-readonly fields, so you can't e.g. edit an `id` or a contact's `display_name`)

The 'FIXME' comments are to say that the solution is imperfect. And honestly I'm not sure I agree we ought to err on the side of restrictive. If we are wrong, then the result is users getting locked-out from something they ought to be allowed to do. OTOH if we are under-restrictive, then the worst that happens is an error message when the user tries to edit something and can't.

The more I think about this, I want to just delete this commit from the PR and rework it separately. I'd like to see this PR merged so it can be beta-tested and not blocked by presentation-layer permissions. I'd rather work up a solution that checks fine-grained ACLs per row, and applies them not only to the edit-in-place fields but also to the links and buttons and menus that might be linking to a screen the user doesn't have permission to see (e.g. Delete Contact).

eileenmcnaughton replied:
@colemanw I guess you are right that it's a usability issue not a security bypass - I'm OK to merge this as is if you want to keep working on it.

I was wanting to see more comments in the actual code - but of course it might be changing again

I'm a bit unsure if we have gone the wrong track not having equivelents to 'edit all contacts' like 'edit all contributions'

colemanw replied:
My thought is to rip out that last commit which checks coarse-grained permissions and instead create a new api action:

```php
// Check if the current user has permission to update participant id 123
civicrm_api4('Participant', 'check', [
  'user_id' => 'current_user', // this would be the default
  'action' => 'update',
  'where' => [['id', '=', 123]], // could potentially check multiple records at once
);
```

This action would check coarse-grained permissions first, and then check ACLs if the entity has any.

This api would give us a consistent way to check permissions for any entity/action combo without actually doing the action (I guess that's pretty similar to @totten's `dry_run` concept).

colemanw commented:
@eileenmcnaughton I've ripped out commit 68da2a3473d4456323b7d87a67f06bfdf5bffac5 (saved it to another branch for posterity but I don't think it's the right solution for having the presentation layer reflect what the user can actually do).

Can we finally merge this?

eileenmcnaughton commented:
@colemanw I'm OK to merge this as long as we try to get something merged into the same (or at worse) next point version since it is pretty confusing to non-permissioned users

colemanw commented:
Permissions were re-implemented via the checkAccess action: https://github.com/civicrm/civicrm-core/pull/20533
